File: compilers/mpi/privatization.cpp
#include <mpi.h>
#include <cstdlib>
#include <iostream>

using namespace std;

double *privatize;

int main( int argc, char **argv )
{
	int rank, size;
	MPI_Status status;
	MPI_Request request;
	int i = 0;
	double *addr;
	MPI_Init( &argc, &argv );
	MPI_Comm_rank( MPI_COMM_WORLD, &rank );
	MPI_Comm_size( MPI_COMM_WORLD, &size );

	//privatization
	for(i=0 ; i<size ; i++)
	{
		if(i != rank)
		{
			double *pointeur = (double *)&privatize;
			MPI_Isend(&pointeur, 1, MPI_DOUBLE, i, 0, MPI_COMM_WORLD, &request);
			MPI_Irecv(&addr, 1, MPI_DOUBLE, i, 0, MPI_COMM_WORLD, &request);

			if(addr == (double *)&privatize)
			{
				cout << "PRIVATIZATION FAILURE\n" << endl;
				abort();
			}
		}
	}
	MPI_Wait(&request, &status);
	MPI_Barrier(MPI_COMM_WORLD);
	if(rank == 0)
		cout << "PRIVATIZATION SUCCESS\n" << endl;
	MPI_Finalize();
	return EXIT_SUCCESS;
}

File: gcc/privatization/fortran/privatization.f90
!~         program main
		subroutine mpc_user_main
        
        include 'mpif.h'
        
!      	Variables globales
        integer privatize
        common /comon/ privatize

        integer ierror
        integer rank

        call mpi_init(ierror)
        call mpi_comm_size(MPI_COMM_WORLD, sizes, ierror)
        call mpi_comm_rank(MPI_COMM_WORLD, rank, ierror)

		privatize = rank;
		call mpi_barrier(MPI_COMM_WORLD, ierror)
		
		if(rank .ne. privatize) then
			write(6,*) 'PRIVATIZATION FAILURE'
			call abort()
		endif
        call mpi_barrier(MPI_COMM_WORLD, ierror)
        
        if(rank .eq. 0) then
             write(6,*) 'PRIVATIZATION SUCCESS'
        endif
        call mpi_finalize(ierror)
        end
